app/src/androidTest/java/com/github/se/project/ui/endToEnd/TutorFindLessonEndToEnd.kt

```diff
@@ -0,0 +1,367 @@
+              status = LessonStatus.STUDENT_REQUESTED,
+              latitude = 46.518973490411526,
+              longitude = 6.5685102716088295),
+      )
+
+  private val mockStudent =
+      Profile(
+          uid = "student123",
+          token = "",
+          googleUid = "mockTutor",
+          firstName = "Ozymandias",
+          lastName = "Halifax",
+          phoneNumber = "1234567890",
+          role = Role.STUDENT,
+          section = Section.IN,
+          academicLevel = AcademicLevel.BA3,
+          languages = listOf(Language.ENGLISH),
+          subjects = listOf(Subject.AICC),
+          schedule = List(7) { List(12) { 0 } })
+
+  @get:Rule val composeTestRule = createComposeRule()
+
+  @get:Rule
+  val grantNotificationPermission: GrantPermissionRule =
+      GrantPermissionRule.grant(android.Manifest.permission.POST_NOTIFICATIONS)
+
+  @get:Rule
+  val permissionRule: GrantPermissionRule =
+      GrantPermissionRule.grant(android.Manifest.permission.ACCESS_FINE_LOCATION)
+
+  @Before
+  fun setUp() {
+    context = mock(Context::class.java)
+    val mockFirebaseMessaging = mock<FirebaseMessaging>()
+    whenever(mockFirebaseMessaging.token).thenReturn(Tasks.forResult("mock_fcm_token"))
+    whenever(mockProfileRepository.addProfile(any(), any(), any())).thenAnswer { invocation ->
+      val onSuccess = invocation.arguments[1] as () -> Unit
+      onSuccess() // Simulate a successful update
+    }
+    whenever(mockProfileRepository.getProfiles(any(), any())).thenAnswer { invocation ->
+      val onSuccess = invocation.arguments[0] as (List<Profile>) -> Unit
+      onSuccess(listOf(mockStudent)) // Simulate a list of profiles with our beloved Ozymandias
+    }
+    whenever(mockProfileRepository.updateToken(any(), any(), any(), any())).thenAnswer { invocation
+      ->
+      val onSuccess = invocation.arguments[2] as () -> Unit
+      onSuccess() // Simulate a successful update
+    }
+    whenever(mockProfileRepository.getNewUid()).thenReturn("mockUid")
+    whenever(mockLessonRepository.updateLesson(any(), any(), any())).thenAnswer { invocation ->
+      mockLessons = listOf(invocation.arguments[0] as Lesson)
+      val onSuccess = invocation.arguments[1] as () -> Unit
+      onSuccess()
+    }
+    whenever(mockLessonRepository.getLessonsForTutor(eq("mockUid"), any(), any())).thenAnswer {
+        invocation ->
+      val onSuccess = invocation.arguments[1] as (List<Lesson>) -> Unit
+      onSuccess(mockLessons)
+    }
+    whenever(mockLessonRepository.getAllRequestedLessons(any(), any())).thenAnswer { invocation ->
+      val onSuccess = invocation.arguments[0] as (List<Lesson>) -> Unit
+      onSuccess(mockLessons)
+    }
+    whenever(mockLessonRepository.updateLesson(any(), any(), any())).thenAnswer { invocation ->
+      mockLessons = listOf(invocation.arguments[0] as Lesson)
+      val onSuccess = invocation.arguments[1] as () -> Unit
+      onSuccess() // Simulate a successful update
+    }
```

Comment:
Really good job for the mocking of all the actions one could made on the profile, lesson and messaging view models.